### ISSUE TITLE
Update typescript-eslint-parser

### ISSUE THREAD

**Issue by bradzacher** [bradzacher/eslint-plugin-typescript#166]
`typescript-eslint-parser` v21 has been released with a number of fixes (such as [support for `visitorKeys` to make traversal easier](https://github.com/eslint/typescript-eslint-parser/issues/532))

`master` is currently on v15, so there could be a number of breaking changes to our rules.
There's a number of places our rule definitions explicitly traverse `typeAnnotation`, which should no longer be required and could cause double the workload / duplicate error reports.

Issues that ***should*** be fixed by this:
- #155
- #152 / #189
- #117 
- #193

Issues that ***might*** be fixed by this:
- #164
- #126

**Comment by macklinu** [comment]
I think this would unblock #117 as well.

**Comment by bradzacher** [comment]
the problem is that `typescript-eslint-parser` isn't actually a dependency/peerDependency for us.
which means people could easily be using it right now!

Luckily, I think that people are pretty lazy with updating the parser.

It's going to be a bit tedious, but yeah - probably a good idea to gradually bump and PR fixes, one version at a time.
Makes it easier to review and track that way.

**Comment by bradzacher** [comment]
@armano2 has done some great work to get the parser dep to v17, and we have merged a number of bugfixes.

I'll spend some time over this long weekend to fix one or two more (and I saw @weirdpattern had his eye on some too).

Unfortunately, the v21 PR is currently blocked by some parser issues, so rather than delaying the fixes until they are resolved, I am thinking we can do a minor bump release on Sunday.

Does that sound like a good idea (@macklinu, @weirdpattern, @j-f1)?
